python/lsst/daf/butler/datastores/chainedDatastore.py

```diff
@@ -0,0 +1,342 @@
+    """
+
+    containerKey = "datastores"
+    """Key to specify where child datastores are configured."""
+
+    @classmethod
+    def setConfigRoot(cls, root, config, full):
+        """Set any filesystem-dependent config options for child Datastores to
+        be appropriate for a new empty repository with the given root.
+
+        Parameters
+        ----------
+        root : `str`
+            Filesystem path to the root of the data repository.
+        config : `Config`
+            A `Config` to update. Only the subset understood by
+            this component will be updated. Will not expand
+            defaults.
+        full : `Config`
+            A complete config with all defaults expanded that can be
+            converted to a `DatastoreConfig`. Read-only and will not be
+            modified by this method.
+            Repository-specific options that should not be obtained
+            from defaults when Butler instances are constructed
+            should be copied from `full` to `Config`.
+        """
+
+        # Extract the part of the config we care about updating
+        datastoreConfig = DatastoreConfig(config, mergeDefaults=False)
+
+        # And the subset of the full config that we can use for reference.
+        # Do not bother with defaults because we are told this already has
+        # them.
+        fullDatastoreConfig = DatastoreConfig(full, mergeDefaults=False)
+
+        # Loop over each datastore config and pass the subsets to the
+        # child datastores to process.
+
+        containerKey = cls.containerKey
+        for idx, (child, fullChild) in enumerate(zip(datastoreConfig[containerKey],
+                                                     fullDatastoreConfig[containerKey])):
+            childConfig = DatastoreConfig(child, mergeDefaults=False)
+            fullChildConfig = DatastoreConfig(fullChild, mergeDefaults=False)
+            datastoreClass = doImport(fullChildConfig["cls"])
+            newroot = "{}/{}_{}".format(root, datastoreClass.__qualname__, idx)
+            datastoreClass.setConfigRoot(newroot, childConfig, fullChildConfig)
+
+            # Reattach to parent
+            datastoreConfig[f"{containerKey}.{idx}"] = childConfig
+
+        # Reattach modified datastore config to parent
+        # If this has a datastore key we attach there, otherwise we assume
+        # this information goes at the top of the config hierarchy.
+        if DatastoreConfig.component in config:
+            config[DatastoreConfig.component] = datastoreConfig
+        else:
+            config.update(datastoreConfig)
+
+        return
+
+    def __init__(self, config, registry=None):
+        super().__init__(config, registry)
+
+        self.storageClassFactory = StorageClassFactory()
+
+        # Name ourselves with the timestamp the datastore
+        # was created.
+        self.name = "{}@{}".format(type(self).__qualname__, time.time())
```

Comment:
If this is timestamp, it's probably not useful as a publicly-visible name; should this be a private attribute?

Reply:
Will this go away when Registry supports ephemeral datastores directly, then?

Also, why does a constraint on `InMemoryDatastore` affect `ChainedDatastore`?

Reply:
Re InMemoryDatastore: Because I got my wires crossed (InMemoryDatastore does need the timestamp to make it unique). You are right that this name is not used directly in registry because the puts are done by the contained datastores (which need unique names). I could just call this `ChainedDatastore` or else call it `ChainedDatastore[datastore1.name, datastore2.name]`

Reply:
Either is fine with me - you probably have a better sense than I do whether we want a more unique and descriptive name here or something more concise.